File: packages/egress-proxy/py/src/steerable_egress_proxy/forward.py
"""Credential-broker forwarding: plain-HTTP in, authenticated HTTPS out.

This module is the W2.2.2 half of the egress proxy. The sandboxed sidecar
points its LLM `baseUrl` at the *http* scheme of the provider host and routes
through this proxy; the proxy terminates the plain-HTTP request, dials the
real upstream over TLS, and injects the credential header. The secret value
lives only in this process — the sidecar (and anything it spawns) never sees
it, which is the entire point of the broker pattern (codex network-proxy
takes the same route).

Fail-closed rules:

- no ``InjectRule`` configured      → non-CONNECT methods stay 405 (v1 behavior)
- absolute-URI host ≠ rule host     → 403
- client sent its own credential    → stripped, never forwarded
- chunked *request* bodies          → 501 (the OpenAI/Anthropic SDKs send
  Content-Length; anything else is out of v1 scope and must fail loud)

One request per client connection (``Connection: close`` both ways): LLM
streaming connections are long-lived anyway, and dropping keep-alive removes
a whole class of request-smuggling surface.
"""

from __future__ import annotations

import asyncio
import logging
import ssl
from dataclasses import dataclass
from urllib.parse import urlsplit

logger = logging.getLogger(__name__)

#: Headers never forwarded to the upstream: hop-by-hop per RFC 9110 §7.6.1,
#: plus any client-supplied credential (the whole reason the broker exists).
_STRIP_REQUEST_HEADERS = frozenset(
    {
        "connection",
        "keep-alive",
        "proxy-connection",
        "proxy-authorization",
        "te",
        "trailer",
        "transfer-encoding",
        "upgrade",
        "expect",
    }
)


@dataclass(frozen=True, slots=True)
class InjectRule:
    """One credential-injection target.

    ``secret`` is the full header value (e.g. ``"Bearer sk-..."``); the CLI
    resolves it from an env var so the secret never appears in argv. v1
    supports exactly one rule per proxy — one provider per sidecar is the
    deployment reality, and a single rule keeps the fail-closed surface
    trivially auditable.
    """

    host: str
    secret: str
    header: str = "Authorization"
    #: Upstream scheme. ``http`` exists for tests and loopback-only upstreams;
    #: production deployments must keep the default ``https``.
    scheme: str = "https"
    port: int | None = None  # default: 443 for https, 80 for http

    def __post_init__(self) -> None:
        if not self.host:
            raise ValueError("InjectRule.host is required")
        if not self.secret:
            raise ValueError("InjectRule.secret is required (fail-closed)")
        if self.scheme not in ("https", "http"):
            raise ValueError(f"InjectRule.scheme must be https|http, got {self.scheme!r}")

    @property
    def upstream_port(self) -> int:
        if self.port is not None:
            return self.port
        return 443 if self.scheme == "https" else 80


@dataclass(frozen=True, slots=True)
class ForwardedRequest:
    """Parsed + rewritten request ready to send to the upstream."""

    method: str
    path: str  # origin-form (path + query)
    head: bytes  # complete rewritten head, CRLF-terminated with blank line
    body_remaining: int  # Content-Length; 0 = no body
    chunked: bool  # True → caller rejects with 501 (v1)


def parse_and_rewrite_request(
    head: bytes, rule: InjectRule
) -> ForwardedRequest | str:
    """Parse the client head and build the upstream request.

    Returns the rewritten request, or a string rejection reason ("403"/"400"/
    "501") — the caller maps it to the status line. Kept pure (no I/O) so the
    rewrite rules are unit-testable without sockets.
    """
    try:
        text = head.decode("iso-8859-1")
    except UnicodeDecodeError:
        return "400"
    lines = text.split("\r\n")
    parts = lines[0].split(" ")
    if len(parts) != 3 or not parts[2].startswith("HTTP/"):
        return "400"
    method, target, _version = parts
    url = urlsplit(target)
    if url.scheme != "http" or not url.hostname:
        # Origin-form or non-http absolute URI: not a forward-proxy request
        # this rule can serve.
        return "403"
    if url.hostname.lower() != rule.host.lower():
        logger.info("deny forward %s (not the inject host)", url.hostname)
        return "403"

    headers: list[tuple[str, str]] = []
    content_length = 0
    chunked = False
    for line in lines[1:]:
        if not line:
            break
        name, sep, value = line.partition(":")
        if not sep:
            return "400"
        lname = name.strip().lower()
        value = value.strip()
        if lname in _STRIP_REQUEST_HEADERS or lname == rule.header.lower():
            continue  # hop-by-hop, Expect, and any client credential
        if lname == "host":
            continue  # rewritten below
        if lname == "content-length":
            try:
                content_length = int(value)
            except ValueError:
                return "400"
            if content_length < 0:
                return "400"
            continue  # re-emitted canonically below
        headers.append((name.strip(), value))
    # Transfer-Encoding was stripped above; detect it from the raw lines so a
    # chunked *request* fails loud instead of desyncing the stream.
    for line in lines[1:]:
        if not line:
            break
        if line.lower().startswith("transfer-encoding:") and "chunked" in line.lower():
            chunked = True

    path = url.path or "/"
    if url.query:
        path = f"{path}?{url.query}"
    out = [f"{method} {path} HTTP/1.1", f"Host: {rule.host}"]
    out.extend(f"{name}: {value}" for name, value in headers)
    if content_length:
        out.append(f"Content-Length: {content_length}")
    out.append(f"{rule.header}: {rule.secret}")
    out.append("Connection: close")
    head_bytes = ("\r\n".join(out) + "\r\n\r\n").encode("iso-8859-1")
    return ForwardedRequest(method, path, head_bytes, content_length, chunked)


async def forward_request(
    reader: asyncio.StreamReader,
    writer: asyncio.StreamWriter,
    head: bytes,
    rule: InjectRule,
    connect_timeout_s: float,
) -> None:
    """Forward one plain-HTTP request to the rule's upstream over TLS.

    Streams the response back as it arrives (SSE needs incremental delivery).
    Returns a status string for logging; raises nothing on the happy path —
    connection errors are answered with a status line and swallowed like the
    CONNECT path's teardown.
    """
    parsed = parse_and_rewrite_request(head, rule)
    if isinstance(parsed, str):
        reasons = {
            "400": (400, "Bad Request"),
            "403": (403, "Forbidden"),
            "501": (501, "Not Implemented"),
        }
        code, reason = reasons[parsed]
        await _reply(writer, code, reason)
        return
    if parsed.chunked:
        await _reply(writer, 501, "Not Implemented")
        return

    ssl_ctx: ssl.SSLContext | None = None
    if rule.scheme == "https":
        ssl_ctx = ssl.create_default_context()
    try:
        upstream_r, upstream_w = await asyncio.wait_for(
            asyncio.open_connection(
                rule.host, rule.upstream_port, ssl=ssl_ctx, server_hostname=rule.host if ssl_ctx else None
            ),
            timeout=connect_timeout_s,
        )
    except (OSError, asyncio.TimeoutError):
        await _reply(writer, 502, "Bad Gateway")
        return

    try:
        upstream_w.write(parsed.head)
        await upstream_w.drain()
        # Request body: exactly Content-Length bytes (chunked rejected above).
        remaining = parsed.body_remaining
        while remaining > 0:
            chunk = await reader.read(min(64 * 1024, remaining))
            if not chunk:
                upstream_w.close()
                return  # client vanished mid-body; nothing useful to send
            upstream_w.write(chunk)
            remaining -= len(chunk)
        await upstream_w.drain()

        await _stream_response(upstream_r, writer)
    finally:
        upstream_w.close()


async def _stream_response(
    upstream_r: asyncio.StreamReader, client_w: asyncio.StreamWriter
) -> None:
    """Forward the upstream response head verbatim, then the body by framing.

    Framing order per RFC 9112: Transfer-Encoding: chunked wins over
    Content-Length; neither means body runs to EOF (we close, so that's
    well-defined for the client).
    """
    head = await _read_upstream_head(upstream_r)
    client_w.write(head)
    await client_w.drain()
    headers = _parse_response_headers(head)
    if headers.get("transfer-encoding", "").lower() == "chunked":
        await _pipe_chunked(upstream_r, client_w)
        return
    length_s = headers.get("content-length")
    if length_s is not None:
        try:
            remaining = int(length_s)
        except ValueError:
            remaining = 0
        while remaining > 0:
            chunk = await upstream_r.read(min(64 * 1024, remaining))
            if not chunk:
                break
            client_w.write(chunk)
            remaining -= len(chunk)
        await client_w.drain()
        return
    # No framing: body to EOF (SSE upstreams that close-delimit).
    while True:
        chunk = await upstream_r.read(64 * 1024)
        if not chunk:
            break
        client_w.write(chunk)
        await client_w.drain()


async def _read_upstream_head(reader: asyncio.StreamReader) -> bytes:
    """Read the response head. `readuntil` keeps any body bytes that arrived
    in the same segment buffered for the framing pipe — a plain `read` would
    over-consume them and corrupt the Content-Length/chunked accounting."""
    try:
        return await reader.readuntil(b"\r\n\r\n")
    except asyncio.IncompleteReadError as exc:
        return exc.partial
    except asyncio.LimitOverrunError:
        # Pathological upstream head; drain what's buffered and forward it —
        # the framing pipe will treat the rest as body and likely fail, which
        # is honest for a broken upstream.
        return await reader.read(64 * 1024)


def _parse_response_headers(head: bytes) -> dict[str, str]:
    try:
        text = head.decode("iso-8859-1")
    except UnicodeDecodeError:
        return {}
    out: dict[str, str] = {}
    for line in text.split("\r\n")[1:]:
        if not line:
            break
        name, sep, value = line.partition(":")
        if sep:
            out[name.strip().lower()] = value.strip()
    return out


async def _pipe_chunked(
    reader: asyncio.StreamReader, writer: asyncio.StreamWriter
) -> None:
    """Forward chunked framing verbatim, flushing per chunk (SSE latency)."""
    while True:
        size_line = await reader.readline()
        if not size_line:
            return
        writer.write(size_line)
        try:
            size = int(size_line.strip().split(b";", 1)[0], 16)
        except ValueError:
            return  # malformed upstream framing; closing is the honest answer
        if size == 0:
            # Trailer section ends at the blank line.
            while True:
                line = await reader.readline()
                if not line:
                    return
                writer.write(line)
                if line == b"\r\n":
                    await writer.drain()
                    return
        remaining = size + 2  # chunk data + trailing CRLF
        while remaining > 0:
            chunk = await reader.read(min(64 * 1024, remaining))
            if not chunk:
                return
            writer.write(chunk)
            remaining -= len(chunk)
        await writer.drain()


async def _reply(writer: asyncio.StreamWriter, code: int, reason: str) -> None:
    writer.write(f"HTTP/1.1 {code} {reason}\r\ncontent-length: 0\r\n\r\n".encode())
    await writer.drain()

File: packages/egress-proxy/rs/src/forward.rs
//! Credential-broker forwarding (Python `steerable_egress_proxy.forward`).

use std::collections::HashSet;
use std::sync::Mutex;
use std::time::Duration;

use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;

use crate::proxy::reply;

const STRIP_REQUEST_HEADERS: &[&str] = &[
    "connection",
    "keep-alive",
    "proxy-connection",
    "proxy-authorization",
    "te",
    "trailer",
    "transfer-encoding",
    "upgrade",
    "expect",
];

static TEE_LOCK: Mutex<()> = Mutex::new(());

#[derive(Clone, Debug)]
pub struct InjectRule {
    pub host: String,
    pub secret: String,
    pub header: String,
    pub scheme: String,
    pub port: Option<u16>,
}

impl InjectRule {
    pub fn new(
        host: impl Into<String>,
        secret: impl Into<String>,
        header: impl Into<String>,
        scheme: impl Into<String>,
        port: Option<u16>,
    ) -> Result<Self, String> {
        let host = host.into();
        let secret = secret.into();
        let header = header.into();
        let scheme = scheme.into();
        if host.is_empty() {
            return Err("InjectRule.host is required".into());
        }
        if secret.is_empty() {
            return Err("InjectRule.secret is required (fail-closed)".into());
        }
        if scheme != "https" && scheme != "http" {
            return Err(format!(
                "InjectRule.scheme must be https|http, got {scheme:?}"
            ));
        }
        Ok(Self {
            host,
            secret,
            header,
            scheme,
            port,
        })
    }

    pub fn upstream_port(&self) -> u16 {
        self.port
            .unwrap_or(if self.scheme == "https" { 443 } else { 80 })
    }
}

#[derive(Debug)]
pub struct ForwardedRequest {
    pub method: String,
    pub path: String,
    pub head: Vec<u8>,
    pub body_remaining: usize,
    pub chunked: bool,
}

pub fn parse_and_rewrite_request(
    head: &[u8],
    rule: &InjectRule,
) -> Result<ForwardedRequest, &'static str> {
    let text = String::from_utf8_lossy(head);
    let mut lines = text.split("\r\n");
    let request_line = lines.next().ok_or("400")?;
    let mut parts = request_line.split(' ');
    let method = parts.next().ok_or("400")?;
    let target = parts.next().ok_or("400")?;
    let version = parts.next().ok_or("400")?;
    if parts.next().is_some() || !version.starts_with("HTTP/") {
        return Err("400");
    }
    let Some((hostname, path)) = split_absolute_http(target) else {
        return Err("403");
    };
    if !hostname.eq_ignore_ascii_case(&rule.host) {
        return Err("403");
    }

    let mut headers: Vec<(String, String)> = Vec::new();
    let mut content_length = 0usize;
    let mut chunked = false;
    let strip: HashSet<&str> = STRIP_REQUEST_HEADERS.iter().copied().collect();
    let inject_header = rule.header.to_ascii_lowercase();
    for line in lines {
        if line.is_empty() {
            break;
        }
        let Some((name, value)) = line.split_once(':') else {
            return Err("400");
        };
        let lname = name.trim().to_ascii_lowercase();
        let value = value.trim();
        if lname == "transfer-encoding" && value.to_ascii_lowercase().contains("chunked") {
            chunked = true;
        }
        if strip.contains(lname.as_str()) || lname == inject_header {
            continue;
        }
        if lname == "host" {
            continue;
        }
        if lname == "content-length" {
            content_length = value.parse().map_err(|_| "400")?;
            continue;
        }
        headers.push((name.trim().to_string(), value.to_string()));
    }

    let mut out = format!("{method} {path} HTTP/1.1\r\nHost: {}\r\n", rule.host);
    for (name, value) in &headers {
        out.push_str(name);
        out.push_str(": ");
        out.push_str(value);
        out.push_str("\r\n");
    }
    if content_length > 0 {
        out.push_str(&format!("Content-Length: {content_length}\r\n"));
    }
    out.push_str(&format!(
        "{}: {}\r\nConnection: close\r\n\r\n",
        rule.header, rule.secret
    ));
    Ok(ForwardedRequest {
        method: method.to_string(),
        path,
        head: out.into_bytes(),
        body_remaining: content_length,
        chunked,
    })
}

fn split_absolute_http(target: &str) -> Option<(String, String)> {
    let rest = target.strip_prefix("http://")?;
    let (authority, path_q) = match rest.split_once('/') {
        Some((authority, rest)) => (authority, format!("/{rest}")),
        None => (rest, "/".to_string()),
    };
    if authority.is_empty() {
        return None;
    }
    let hostname = if authority.starts_with('[') {
        let end = authority.find(']')?;
        authority[1..end].to_string()
    } else {
        authority
            .rsplit_once(':')
            .filter(|(_, port)| port.chars().all(|c| c.is_ascii_digit()) && !port.is_empty())
            .map(|(host, _)| host.to_string())
            .unwrap_or_else(|| authority.to_string())
    };
    if hostname.is_empty() {
        return None;
    }
    Some((hostname, path_q))
}

pub fn tee_forwarded_request(path: &str, parsed: &ForwardedRequest, body: &[u8]) {
    let record = serde_json::json!({
        "method": parsed.method,
        "path": parsed.path,
        "body_bytes": body.len(),
        "body": String::from_utf8_lossy(body),
    });
    let _guard = TEE_LOCK.lock().unwrap();
    if let Some(parent) = std::path::Path::new(path).parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    use std::io::Write;
    if let Ok(mut file) = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
    {
        let _ = writeln!(file, "{record}");
    }
}

pub async fn forward_request(
    client: &mut TcpStream,
    head: &[u8],
    rule: &InjectRule,
    timeout: Duration,
    record_requests: Option<&str>,
) -> std::io::Result<()> {
    let parsed = match parse_and_rewrite_request(head, rule) {
        Ok(parsed) if parsed.chunked => {
            reply(client, 501, "Not Implemented").await?;
            return Ok(());
        }
        Ok(parsed) => parsed,
        Err("403") => {
            reply(client, 403, "Forbidden").await?;
            return Ok(());
        }
        Err("501") => {
            reply(client, 501, "Not Implemented").await?;
            return Ok(());
        }
        Err(_) => {
            reply(client, 400, "Bad Request").await?;
            return Ok(());
        }
    };

    let mut upstream = match dial_upstream(rule, timeout).await {
        Ok(stream) => stream,
        Err(_) => {
            reply(client, 502, "Bad Gateway").await?;
            return Ok(());
        }
    };
    upstream.write_all(&parsed.head).await?;
    let mut remaining = parsed.body_remaining;
    let mut body_parts = Vec::new();
    let mut buf = vec![0u8; 64 * 1024];
    while remaining > 0 {
        let take = remaining.min(buf.len());
        let n = client.read(&mut buf[..take]).await?;
        if n == 0 {
            return Ok(());
        }
        upstream.write_all(&buf[..n]).await?;
        if record_requests.is_some() {
            body_parts.extend_from_slice(&buf[..n]);
        }
        remaining -= n;
    }
    if let Some(path) = record_requests {
        tee_forwarded_request(path, &parsed, &body_parts);
    }
    let _ = tokio::io::copy(&mut upstream, client).await;
    Ok(())
}

async fn dial_upstream(rule: &InjectRule, timeout: Duration) -> std::io::Result<Upstream> {
    let tcp = tokio::time::timeout(
        timeout,
        TcpStream::connect((rule.host.as_str(), rule.upstream_port())),
    )
    .await
    .map_err(|_| std::io::Error::new(std::io::ErrorKind::TimedOut, "dial"))?
    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
    if rule.scheme == "http" {
        return Ok(Upstream::Plain(tcp));
    }
    let connector = tokio_native_tls::TlsConnector::from(
        native_tls::TlsConnector::new()
            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
    );
    let tls = connector
        .connect(&rule.host, tcp)
        .await
        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
    Ok(Upstream::Tls(tls))
}

enum Upstream {
    Plain(TcpStream),
    Tls(tokio_native_tls::TlsStream<TcpStream>),
}

impl tokio::io::AsyncRead for Upstream {
    fn poll_read(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
        buf: &mut tokio::io::ReadBuf<'_>,
    ) -> std::task::Poll<std::io::Result<()>> {
        match self.get_mut() {
            Upstream::Plain(stream) => std::pin::Pin::new(stream).poll_read(cx, buf),
            Upstream::Tls(stream) => std::pin::Pin::new(stream).poll_read(cx, buf),
        }
    }
}

impl tokio::io::AsyncWrite for Upstream {
    fn poll_write(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
        buf: &[u8],
    ) -> std::task::Poll<std::io::Result<usize>> {
        match self.get_mut() {
            Upstream::Plain(stream) => std::pin::Pin::new(stream).poll_write(cx, buf),
            Upstream::Tls(stream) => std::pin::Pin::new(stream).poll_write(cx, buf),
        }
    }

    fn poll_flush(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<std::io::Result<()>> {
        match self.get_mut() {
            Upstream::Plain(stream) => std::pin::Pin::new(stream).poll_flush(cx),
            Upstream::Tls(stream) => std::pin::Pin::new(stream).poll_flush(cx),
        }
    }

    fn poll_shutdown(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<std::io::Result<()>> {
        match self.get_mut() {
            Upstream::Plain(stream) => std::pin::Pin::new(stream).poll_shutdown(cx),
            Upstream::Tls(stream) => std::pin::Pin::new(stream).poll_shutdown(cx),
        }
    }
}
